codescoop integration

sweetca opened the pull request:
Signed-off-by: Alex Skorohod <alexey.skorohod@gmail.com>

Codescoop integration regarding project issues on 
https://github.com/codescoopltd/cs360/projects/1

sweetca commented:
@alexbrdn @maxhbr @mcjaeger 
this is signed PR for integration part

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/components/detail.jsp:
```diff
@@ -35,6 +36,13 @@
+    <portlet:param name="<%=PortalConstants.COMPONENT_ID%>" value="${component.id}"/>
+</portlet:actionURL>
+
+<c:set var="autocompleteUrl" value="<%=PortalConstants.CODESCOOP_URL%>"/>
```
`autocompleteUrl` is a rather generic name for such a spezific variable

sweetca replied:
renamed

sweetca replied:
@maxhbr unfortunately I can not escape from s:set var in jsp eval...

maxhbr commented on frontend/sw360-portlet/src/main/webapp/js/modules/codeScoop.js:
```diff
@@ -0,0 +1,966 @@
+/*
+ * Copyright Siemens AG, 2017.
```
`Copyright Siemens AG, 2017. ` is correct?

sweetca replied:
@maxhbr which copyright I should put here?

mcjaeger replied:
I will clarify

sweetca replied:
@mcjaeger Michael, do you have update for this issue? can i use such copyright?

mcjaeger replied:
for entirely new files like this please use the copyright from your organisation.

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/components/view.jsp:
```diff
@@ -238,6 +265,43 @@
+                var order = [[1, 'asc']];
+                var columnDefs =  [];
+
+                if (window.codescoopEnabled) {
```
Why is't this handled in jsp?

maxhbr replied:
This looks like hard to maintain code. E.g. hard to test whether the column widths and styles are still good if no CS support is present.

sweetca replied:
@maxhbr regarding @mcjaeger request to deal with table UI we need to catch it here to make table init process match to column size

its totally easy to see same result as before if codescoop is not enabled: page will be same as now on your instances

mcjaeger replied:
I think the issue is that if someone now changes the table, without having access to codescoop, there is no chance of testing if the changes break the codescoop view.

sweetca replied:
updated

sweetca replied:
@mcjaeger how we should proceed here?

mcjaeger replied:
@maxhbr what would be alternatives here? I think some logic is necessary in JavaScript as the altering of the table.

maxhbr replied:
One could move the logic to the jsp level, e.g. by `<core_rt:if test="${not empty CODESCOOP_URL && not empty CODESCOOP_TOKEN}">`.

maxhbr commented on frontend/sw360-portlet/src/main/webapp/html/components/includes/releases/editReleaseInformation.jspf:
```diff
@@ -146,3 +146,16 @@
 	});
 </script>
+
+<core_rt:if test="${not empty autocompleteUrl && not empty autocompleteToken}">
```
Why is `PortalConstants.CODESCOOP_URL` mapped to `autocompleteUrl` to be used here, instead of using `PortalConstants.CODESCOOP_URL` directly in some way here?

sweetca replied:
@maxhbr general idea was that if some one wants do not destroy component data on portion update  you may use it like trigger for backend 
see https://github.com/eclipse/sw360/pull/300 where I provided fix
for now was decision  to escape from constant - will remove also from JSP

alexbrdn replied:
This answer seems misplaced. I don't think it addresses @maxhbr's question.

sweetca replied:
refactored

maxhbr commented on frontend/sw360-portlet/src/main/webapp/js/modules/codeScoop.js:
```diff
@@ -0,0 +1,966 @@
+        modalBox.style.zIndex = "999999";
+        modalBox.style.border = "1px solid #c5c5c5";
+        modalBox.style.backgroundColor = "#ffffff";
+        modalBox.style.color = "#333333";
```
Is this code generated? There are many magic strings, which could be moved to constants.

sweetca replied:
@maxhbr basic logic that I do not want put them as a constants for now. this solution will still going in development for UI look&feel
as first iteration will have no any dependencies from sw360 css variables and will not be damaged by extra changes in sw360 css

mcjaeger replied:
as I understood the issue, @maxhbr is pointing to this issue: https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad

sweetca replied:
@mcjaeger this is css hex color that sw360 have in codebase also. nothing magic here: just init css directly from js code without inheritance of sw360 basic styles to save my code from css conflicts

mcjaeger replied:
(from discussion)
* just move the values to the const values to the beginning of that file.  (preferred)
* one solution is to create a separate JS object that contains all the style information (rather than having a css file (ok)
* separate css file which is either loaded all the times at page load or nastily loading a separate css at javascript runtime (not so good)

maxhbr commented on frontend/sw360-portlet/src/main/webapp/js/modules/codeScoop.js:
```diff
@@ -0,0 +1,966 @@
+        };
+
+        this.activateIndexes = function(tableID, dataUrl) {
+            dataUrl = dataUrl + "&sEcho=1&iColumns=5&sColumns=%2C%2C%2C%2C&iDisplayStart=0" +
```
This long hardcoded string looks fragile

sweetca replied:
@maxhbr this is part constant for all kind request...
is it better solution to make this works?

mcjaeger replied:
one solution could using the builder pattern to build an URL. i am not sure if this works straight away in javascript though. For java there are some implementations for this.

sweetca replied:
@mcjaeger unfortunately I can not pass to such build data from JS

alexbrdn commented on frontend/sw360-portlet/src/main/webapp/js/modules/codeScoop.js:
```diff
@@ -0,0 +1,966 @@
+            var xhr = new XMLHttpRequest();
+            xhr.open(method, _this.apiUrl + path, true);
+            xhr.setRequestHeader("X-Api-Key", _this.apiToken);
+            xhr.setRequestHeader("X-User-Login", "test");
```
is this a hard-coded user name?

sweetca replied:
@alexbrdn its hardcoded access type for CD system. sw360 will not send any type of personal data from this script

alexbrdn replied:
It's not the personal data I have concerns for. I was wondering whether this is some kind of parameter that should be parametrized rather than hard-coded. When you hard-code `"test"` it does not look like it's meant for production, you know.

sweetca commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortletUtils.java:
```diff
@@ -123,21 +124,23 @@ private static Repository getRepositoryFromRequest(PortletRequest request) {
+        for (Component._Fields field : extractFieldsForComponentUpdate(requestParams, component)) {
+            setFieldValue(request, component, field);
         }
+        if (!requestParams.contains(PortalConstants.UPDATE_ONLY_REQUESTED)) {
```
@alexbrdn right. will escape this after merge of https://github.com/eclipse/sw360/pull/300 and rebasing this PR on top of master

sweetca replied:
removed from current PR

sweetca commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortletUtils.java:
```diff
@@ -123,21 +124,23 @@ private static Repository getRepositoryFromRequest(PortletRequest request) {
+        return component.getId() == null
+                ? Arrays.asList(Component._Fields.values())
+                : Arrays.stream(Component._Fields.values())
+                .filter(f -> requestParams.contains(f.name()))
```
@alexbrdn I'm trying to catch only relational fields that can be damaged if you posting portion of entity...

sweetca replied:
removed from current PR

sweetca commented on frontend/sw360-portlet/src/main/resources/sw360.properties:
```diff
@@ -44,3 +44,8 @@
+
+## Codescoop integration
+codescoop.url=https://api.codescoop.com/
+codescoop.token=db28adb9-87f1-4cfa-b3ae-c00443252d8d
```
@alexbrdn its a demo token that will be died after. before merge PR this will be commented out & codescoop will be disabled by default

sweetca replied:
@alexbrdn now this file updated

sweetca commented on frontend/sw360-portlet/src/main/webapp/css/dataTable_Siemens.css:
```diff
@@ -171,6 +171,7 @@ table.display tr.heading2 td {
 
 table.display td {
 	padding: 3px 10px;
+	min-width: 30px;
```
@alexbrdn its a bugfix by request from @mcjaeger

mcjaeger replied:
actually, I am not sure, pls help to point out to which Bugfix you are referring to?

But I am not OK with the naming of the file. There should not be files named "Siemens" or so in the Eclipse project.

sweetca replied:
@mcjaeger regarding renaming this file: I'm not owner of it so cant put correct name there.
this is css fix that solve troubles when you compress browser width to minimal. in this case cells are start compressing content even to 0 width. this fix does not destroy cell width and content will save shape...

sweetca commented on frontend/sw360-portlet/src/main/webapp/html/components/edit.jsp:
```diff
@@ -141,93 +155,94 @@
-
-        $('#formSubmit').click(
-            function () {
+    require(['jquery',
```
ok

sweetca commented on frontend/sw360-portlet/src/main/webapp/html/components/includes/components/editBasicInfo.jspf:
```diff
@@ -136,6 +138,8 @@
             checkComponentNameParamKey = '<portlet:namespace/><%=PortalConstants.COMPONENT_NAME%>',
             openOnBlurOfElementId = '#comp_name';
 
+        <core_rt:if test="${empty autocompleteUrl && empty autocompleteToken}">
```
@alexbrdn you have wrong feelings here

sweetca replied:
I want inject code only if 2 properties are enabled
I want to check exact not empty for bought of them...

sweetca replied:
@alexbrdn you right, thanks for catch.
updated...

alexbrdn commented:
The changes I requested are fine by me. 
@sweetca, please make sure to update that questionable copyright statement and clarify the points raised by @maxhbr.

@maxhbr, if you could give an approved review to unambiguously indicate that you have no objections to merging, that'd be great.

maxhbr commented:
> @maxhbr, if you could give an approved review to unambiguously indicate that you have no objections to merging, that'd be great.

I have not yet reviewed the PR completly. I have just scrolled through the code.

sweetca commented:
rebased on top of eclipse master

sweetca commented:
@alexbrdn @maxhbr @mcjaeger 
refactored, rebased on top of latest master
Changes:
- copyrights to Codescoop Oy
- jsp fix for CD plugin in disabled mode
- js codestyle fix

sweetca commented:
@mcjaeger @alexbrdn @maxhbr css updated
